### ISSUE TITLE
Bert backward propagation order isn't correct ??

### ISSUE THREAD

**Issue by YouhuiBai** [dmlc/gluon-nlp#861]
## Description
Hi, I downloaded Bert source code from [Bert](https://gluon-nlp.mxnet.io/model_zoo/bert/index.html), and ran successfully at a single machine with one GPU. Then I expanded the code and made it be with the support of Horovod, modified source code [here](https://gitee.com/YouhuiBai/Bert), only changed the `finetune_classifier.py`, then I ran it with 2 servers with one GPU per machine, and profiled the process with [Horovod timeline](https://github.com/horovod/horovod/blob/master/docs/timeline.rst), found that there was no overlap between backward computation and communication, namely, it begins to transfer after the completion of backward, cause the order of transfer is exactly the same with forward computation order.

The earliest operators completed from mxnet engine will be transferred first by Horovod, but the transfer order is the same with forward's order, so I think the backward completion order from mxnet engine is not correct. Can anyone give some suggestions? 

Thanks.

## To Reproduce
I used the latest code from [Bert](https://gluon-nlp.mxnet.io/model_zoo/bert/index.html) and made it be with the support of horovod too, the same result I can get.

## What have you tried to solve it?

1. I changed the Bert computation graph from static to dynamic, but it didn't matter.
2. I tracked to mxnet imperative computation graph, the node ids indicate that the backward order is the reverse of forward, then the operators are sent to mxnet engine to execute with dependency.

But the Bert from google-research with tensorflow and horovod does have overlap between backward and communication. And the image-classification models from [here](https://github.com/horovod/horovod/blob/master/examples/mxnet_imagenet_resnet50.py) also have overlap with mxnet and horovod.

## Environment

gluonnlp 0.7.1, mxnet 1.5.0, horovod 0.16.1, openmpi 3.1.2

# paste outputs here
For example, you can see that the transformer1 is transferred after transformer0, but the transformer1 is computed and completed first in backward propagation in theory.
![image](https://user-images.githubusercontent.com/27176645/62412480-886ede80-b635-11e9-8583-6ab603bb55d3.png)

**Comment by eric-haibin-lin** [comment]
This looks like a bug. Would you mind providing the complete hvd timeline log?

**Comment by YouhuiBai** [comment]
@eric-haibin-lin Of course, the log is [Horovod timeline for Bert](https://github.com/dmlc/gluon-nlp/files/3464459/timeline_Bert.txt),  you can modify the extension name from `.txt` to `.json` to open  with chrome.

The setups are:
1. Bert using RTE dataset and running 50 iterations
2. Single machine with 2 GPUs, the communication way between GPUs is PCIe, Horovod organizes these two GPU as a ring

You can see that from the first message prepared to the last message, it takes about 30 ms per iteration, means that the backward computation time cost is less than 30 ms, it's unreasonable.

Thanks a lot

**Comment by YouhuiBai** [comment]
@eric-haibin-lin Hi Haibin, any suggestions?

**Comment by eric-haibin-lin** [comment]
Hi @YouhuiBai can you update your horovod version? It was fixed by this PR: https://github.com/horovod/horovod/commit/0febb667bd0f511de61d2f2d7a667d8e9d30f821 
where the allreduce operation is scheduled by priority 
Although the priority currently is defined by the parameter name, which does not really reflect the dependency in a general graph..

**Comment by YouhuiBai** [comment]
@eric-haibin-lin Thank you,  I see, this is a way to fix this problem by pass a prioritized parameter to mxnet engine, the engine will schedule the operations depended on variables by priority, but why is the keys' order out from the engine different from the order of backward propagation when there is no priority? I think the problem lies in the default scheduling method of mxnet engine, and try to solve it, if you have some experience, please share to me.

**Comment by eric-haibin-lin** [comment]
The idea is that we want to schedule layers that are closer to the input earlier, if when multiple tensors are available for allreduce. This paper talks about the priority based approach: https://www.sysml.cc/doc/2019/75.pdf 
Are you trying to improve the allreduce strategy in mxnet/horovod?

**Comment by YouhuiBai** [comment]
Hi Haibin, thanks and yes, I am trying to improve the allreduce strategy in mxnet with horovod, I tried to analyze the computation and communication order of image-classification models and NLP models in mxnet+horovod, then found that the transferring order for Bert is strange. I have some suspects and am trying to verify them these days.

**Comment by eric-haibin-lin** [comment]
I think there's a bug with the newly added sort logic: https://github.com/horovod/horovod/blob/3999eca9c85550564444c61db55591e86c584ea6/horovod/mxnet/__init__.py#L101-L102 @YouhuiBai

**Comment by YouhuiBai** [comment]
@eric-haibin-lin Sorry I forgot to check gmail last days, I used horovod with version 0.16.1, there is no `sort` in `_allreduce_grads()`. I found that this is because of global gradient clipping, namely, when all gradients are completed then compute the clipping `total_norm` in `clip_grad_global_norm()`, if the clipping is followed by communication, of course there is no overlap between backward computation and communication. And it doesn't matter to the priority in Mxnet engine, the priority will only impact the order of operators in `task_queue` at the same moment. Thanks a lot!!
